Make `FittedBox` not throw when child has zero size.

PurplePolyhedron opened the pull request:
fix https://github.com/flutter/flutter/issues/135082 , fix https://github.com/flutter/flutter/issues/139539 , fix https://github.com/flutter/flutter/issues/142910

Before, `FittedBox` would throw when child size is zero, unless the constraint is tight and fit is not `BoxFit.scaleDown`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

PurplePolyhedron committed:
Update fitted_box layout

PurplePolyhedron committed:
Update fitted_box_test

PurplePolyhedron committed:
Update fitted_box_test

PurplePolyhedron committed:
Update fitted_box_test

goderbauer commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -2729,18 +2729,23 @@ class RenderFittedBox extends RenderProxyBox {
-        case BoxFit.none:
-          size = constraints.constrainSizeAndAttemptToPreserveAspectRatio(child!.size);
+      if (child!.size.isEmpty) {
+        size = constraints.biggest;
```
Our default is usually `constraints.smallest` for these cases (see also the else branch bellow). Was there a good reason to use biggest here? Or should this also be smallest?

PurplePolyhedron replied:
My bad, I made some mistake in previous testing and thought the default behaviour is `constraints.smallest`

PurplePolyhedron replied:
I've changed it to smallest. And I updated `computeDryLayout` to match `performLayout` since I forgot about `computeDryLayout` perviously.

goderbauer reviewed (commented):
Thank you for the fix and sorry for the delay in reviewing it.

PurplePolyhedron committed:
Update proxy_box

change fitted box default size to constraint.smallest and match computeDryLayout to performLayout

goderbauer reviewed (approved):
LGTM!

LongCatIsLooong commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -2729,18 +2715,23 @@ class RenderFittedBox extends RenderProxyBox {
-        case BoxFit.none:
-          size = constraints.constrainSizeAndAttemptToPreserveAspectRatio(child!.size);
+      if (child!.size.isEmpty) {
+        size = constraints.smallest;
```
The assertion was in `BoxConstraints.constrainSizeAndAttemptToPreserveAspectRatio`. I think we should try to fix that method so it can handle an empty size as input?

PurplePolyhedron replied:
I am not sure how it should behave when only one side of the target size is 0.  For example if `height` is finite while `width` is 0, Do you think `BoxConstraints.constrainSizeAndAttemptToPreserveAspectRatio` should trying to match the height(only constrain size since the aspect ratio is 0 and cannot be Preserved), or return `constraints.smallest` regardless(treat any zero size as empty)?

LongCatIsLooong replied:
Yeah I think in your example the method should return `constrainSize(inputSize)`, instead of returning `smallest` and ignoring the input size.

The reason for that is I think returning `smallest` when the child size is empty feels like a special case, one that feels a bit artificial (if we do that then the documentation needs to be updated). 
Say the input `BoxConstraints` is loose and unbounded (i.e., `BoxConstraints()`), then `RenderFittedBox` always takes the same size as its child when the child size is not empty, so ideally we don't have the special case it when the child size is empty, and always return the child size regardless if it's empty.

PurplePolyhedron committed:
Merge branch 'flutter:master' into fittedbox_size

PurplePolyhedron committed:
revert proxy_box

PurplePolyhedron committed:
Update constrainSizeAndAttemptToPreserveAspectRatio

PurplePolyhedron committed:
Update constrainSizeAndAttemptToPreserveAspectRatio test

PurplePolyhedron committed:
fix typo

PurplePolyhedron committed:
Update RenderFittedBox.computeDryLayout

remove guard in RenderFittedBox.computeDryLayout since zero size child is now valid.

LongCatIsLooong commented:
@goderbauer would you like to take another look and see if the updated version makes sense to you?

goderbauer reviewed (approved):
Still LGTM

LongCatIsLooong reviewed (approved):
Thank you for contributing!
